src/ttsim/interface_dag_elements/data_converters.py

```diff
@@ -164,10 +172,19 @@ def df_with_nested_columns_to_flat_data(
         >>> result
         {("a", "b"): np.array([1, 2, 3]), ("c",): np.array([4, 5, 6])}
     """
-    return {
-        _remove_nan_from_keys(key): xnp.asarray(value)
-        for key, value in df.to_dict(orient="list").items()
-    }
+    # Always use numpy for initial data preparation (performance optimization for JAX)
+    result = {}
+    for key, value in df.to_dict(orient="list").items():
+        clean_key = _remove_nan_from_keys(key)
+        numpy_array = np.asarray(value)
+
+        # Convert to target backend array with zero-copy if JAX
+        if xnp.__name__ == "jax.numpy":
+            result[clean_key] = xnp.asarray(numpy_array, copy=False)
```

Comment:
```suggestion
            result[clean_key] = xnp.asarray(numpy.asarray(value)
, copy=False)
```
Here I am not sure whether the `copy=False` works when we are using the GPU backend. @mj023, could you try that out?

Reply:
The [Jax documentation](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.asarray.html) states:

> - copy ([bool](https://docs.python.org/3/library/functions.html#bool) | None) – optional boolean specifying the copy mode. If True, then always return a copy. If False, then error if a copy is necessary. Default is None, which will only copy when necessary.

If you take that literally, this seems to suggest that `xnp.asarray(numpy_array, copy=False)` will error with the JAX-GPU backend because copying of the NumPy array to the GPU is definitely necessary. However, I find it quite likely that JAX will simply ignore the parameter when CPU->GPU transfer is necessary. You cannot _not_ copy in that case! But yes, impossible to know for sure without trying it out...

Reply:
You are right, it does not work with the GPU backend, I just tested it. But that is not a problem, it still works the same, if you just remove `copy=False`. The speedup seems to come from first converting to numpy arrays, this [here](https://github.com/jax-ml/jax/issues/10662) might explain it.

Reply:
You are also right, the performance hit of removing `copy=False` is almost non-existent. I was more concerned about memory usage. The input data might be many GBs large, so we don't want to make unnecessary copies.

I just added (okay, Claude did) peak memory usage to `benchmark.py`: 
[benchmark.zip](https://github.com/user-attachments/files/21669057/benchmark.zip)

### WITHOUT `copy=False`
```
=======================================================================================================================================
MEMORY USAGE COMPARISON
=======================================================================================================================================
Households  Baseline    Data Size   NumPy Peak  JAX Peak    NumPy Inc   JAX Inc     NumPy Ratio JAX Ratio   Peak Ratio
(MB)        (MB)        (MB)        (MB)        (MB)        (MB)        (MB)        (Inc/Data)  (Inc/Data)  (JAX/NumPy)
---------------------------------------------------------------------------------------------------------------------------------------
32,767      133.1       42.5        186.7       269.3       11.1        91.2        0.3x        2.2x        1.44x
32,768      136.7       40.3        187.9       295.8       10.9        40.1        0.3x        1.0x        1.57x
65,536      137.3       82.0        240.7       371.7       21.3        45.1        0.3x        0.6x        1.54x
131,072     135.7       164.3       340.4       503.4       40.4        68.8        0.2x        0.4x        1.48x
262,144     135.6       327.1       543.8       735.6       81.1        113.2       0.2x        0.3x        1.35x
524,288     136.1       654.1       951.0       1179.7      160.8       209.2       0.2x        0.3x        1.24x
1,048,576   136.4       1308.1      1777.7      2047.5      333.3       397.9       0.3x        0.3x        1.15x
2,097,152   136.4       2616.1      3462.7      3654.3      710.2       666.9       0.3x        0.3x        1.06x
```

### WITH `copy=False`
```
=======================================================================================================================================
MEMORY USAGE COMPARISON
=======================================================================================================================================
Households  Baseline    Data Size   NumPy Peak  JAX Peak    NumPy Inc   JAX Inc     NumPy Ratio JAX Ratio   Peak Ratio
(MB)        (MB)        (MB)        (MB)        (MB)        (MB)        (MB)        (Inc/Data)  (Inc/Data)  (JAX/NumPy)
---------------------------------------------------------------------------------------------------------------------------------------
32,767      133.0       42.5        186.5       269.0       10.9        90.9        0.3x        2.2x        1.44x
32,768      136.5       40.4        187.3       294.4       10.3        37.1        0.3x        0.9x        1.57x
65,536      137.3       80.7        240.1       375.5       22.1        49.9        0.3x        0.6x        1.56x
131,072     135.4       164.6       344.2       507.0       44.2        70.4        0.3x        0.4x        1.47x
262,144     135.6       327.1       547.0       740.6       84.3        114.6       0.3x        0.4x        1.35x
524,288     136.1       654.1       966.7       1181.0      176.5       205.5       0.3x        0.3x        1.22x
1,048,576   136.7       1308.1      1790.4      2045.3      345.6       394.4       0.3x        0.3x        1.14x
2,097,152   136.8       2616.1      3457.9      3741.8      705.0       756.6       0.3x        0.3x        1.08x

Note: Peak Ratio > 1.0 means JAX uses more peak memory than NumPy
Note: 'Inc' = Memory increase during execution (peak - start)
Note: 'Ratio' = Memory increase relative to input data size
Note: Higher ratios indicate more memory overhead relative to input data
```

It seems `copy=False` doesn't really help very much here. 😅 So my worries were very much unfounded and it is basically costless to remove `copy=False` as you suggested.